Add k8s container logs collection

astencel-sumo opened the pull request:
Fixes https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/2536

naseemkullah commented on charts/opentelemetry-collector/README.md:
```diff
@@ -107,6 +109,44 @@ agentCollector:
+
+This feature needs agent collector to be deployed, which means it will not work if only standalone collector is enabled.
+
+The feature can be disabled by setting the `agentCollector.containerLogs.enabled` property to `false`.
```
For metrics and traces pipelines, we set them to null to disable, would it be possible to keep this pattern?

```yaml
config:
  service:
    pipelines:
      logs: null
```

astencel-sumo replied:
It seems to me these are two different things: disabling the whole logs pipeline vs. disabling container logs collection. You might want to disable container logs collection but keep the logs pipeline for other uses?

astencel-sumo replied:
Unless the container logs collection should be in its own pipeline? As opposed to a general "logs" pipeline.

naseemkullah replied:
a `logs` pipeline does seem like the right thing

naseemkullah commented on charts/opentelemetry-collector/values.yaml:
```diff
@@ -63,17 +71,17 @@ config:
-  repository: otel/opentelemetry-collector
-  # repository: otel/opentelemetry-collector-contrib
+  # repository: otel/opentelemetry-collector
+  repository: otel/opentelemetry-collector-contrib
```
will logs eventually land in core? / is the contrib image a better default?

sumo-drosiek replied:
filelog receiver which is going to be used for scrapping logs from kubernetes env is under development in contrib. I think that it is not going to be included in core soon if anytime

naseemkullah replied:
I see, thanks

tigrannajaryan replied:
@dmitryax any thoughts on what the implications of these change could be? Contrib is a much larger image, do we care about it?

dmitryax replied:
The original idea for this chart was is to support only the core collector (like "essential") functionality by default and provide users with an option to use the contrib collector for other use cases. If we can move `filelog` receiver to the core collector eventually, I'm good with temporarily switching it to contrib.

dmitryax replied:
If the logging pipeline is unstable, I'd recommend to set `containerLogs.enabled` to `false` by default, and use contrib build only if user sets it to `true`. @tigrannajaryan @naseemkullah thoughts?

naseemkullah replied:
sgtm, provided the user can still override the image in either case

tigrannajaryan replied:
> If the logging pipeline is unstable, I'd recommend to set `containerLogs.enabled` to `false` by default, and use contrib build only if user sets it to `true`. @tigrannajaryan @naseemkullah thoughts?

@dmitryax Is it possible use different builds/images based on a chart setting? I think that's not a bad idea to do temporarily until filelog is moved to the core.

dmitryax replied:
> @dmitryax Is it possible use different builds/images based on a chart setting?

It can be a bit tricky, but must be possible by introducing another template for the image and command arguments.

>  I think that's not a bad idea to do temporarily until filelog is moved to the core.

If we will eventually move the `filelog` receiver to the core, I'm good with temporarily bringing contrib image. Let's put a comment saying that if user don't need the logs pipeline they can set the image back to core.

tigrannajaryan replied:
The exact set of receivers we will have in the core is currently being discussed. We are not yet sure the full `filelog` will be in the core. Let's assume that maybe it will not happen in the near future. We need a reasonable solution for now that uses `filelog` from contrib.

dmitryax replied:
let's see if we can easily image the image (core vs contrib) dependant on `containerLogs.enabled`. if it's becomes too involved, we can keep contrib image by default?

perk-sumo replied:
> I'd recommend to set `containerLogs.enabled` to `false` by default, and use `contrib` build only if user sets it to true

Hi!
The `containerLogs.enabled=false` by default make sure at this point.
But I think this should change pretty soon, as soon as we confirm this is working.
That's because gathering logs from the k8s environment is one of the most obvious things a user would want from this chart IMO.

naseemkullah replied:
While that may be true for other users, today as a user of this chart in production I would disable k8s container logging as that is covered by my providers managed service, I use the collector exclusively for trace data collection.

dmitryax replied:
I agree that we shouldn't enable any functionality that isn't part of the stable core functionality

astencel-sumo replied:
I have disabled container logs by default and reverted to using core collector image in https://github.com/open-telemetry/opentelemetry-helm-charts/pull/36/commits/f005dde14bfdd2971aad69a3c0adbe3233a78a97.

naseemkullah commented on charts/opentelemetry-collector/values.yaml:
```diff
@@ -39,6 +39,14 @@ config:
     pipelines:
+      logs:
+        exporters:
+          logging
```
```suggestion
          - logging
```

naseemkullah commented:
PTAL @dmitryax

tigrannajaryan commented on charts/opentelemetry-collector/templates/_config.tpl:
```diff
@@ -115,9 +116,89 @@ exporters:
+receivers:
+  filelog:
+    include: [ /var/log/pods/*/*/*.log ]
+    exclude: [ /var/log/pods/{{ .Release.Namespace }}_{{ include "opentelemetry-collector.fullname" . }}*_*/{{ .Chart.Name }}/*.log ]
```
Please add a comment to explain what this does. I believe we need to make this warning very clear since it can be dangerous to collect own logs.

astencel-sumo replied:
I have added a comment in https://github.com/open-telemetry/opentelemetry-helm-charts/pull/36/commits/4df897e8729399eaa3e4fde019ca7a3b31d9a316 and more documentation in https://github.com/open-telemetry/opentelemetry-helm-charts/pull/36/commits/f005dde14bfdd2971aad69a3c0adbe3233a78a97. Please take a look @tigrannajaryan.

tigrannajaryan commented on charts/opentelemetry-collector/values.yaml:
```diff
@@ -63,17 +71,17 @@ config:
 image:
-  repository: otel/opentelemetry-collector
-  # repository: otel/opentelemetry-collector-contrib
+  # repository: otel/opentelemetry-collector
```
Remove commented line.

astencel-sumo replied:
Done

tigrannajaryan commented on charts/opentelemetry-collector/values.yaml:
```diff
@@ -63,17 +71,17 @@ config:
 command:
-  name: otelcol
-  # name: otelcontribcol
+  # name: otelcol
```
Remove commented line.

astencel-sumo replied:
Done

astencel-sumo commented on charts/opentelemetry-collector/README.md:
```diff
@@ -107,6 +109,44 @@ agentCollector:
+To change that, replace the exporter on the logs pipeline and remove the exclude on the logs receiver.
+
+Here's an example `values.yaml` file that replaces the default `logging` exporter on the `logs` pipeline
+with a `file` exporter that writes the container logs to `/var/log/container-logs/container-logs.json`.
```
> I believe using file exporter AND collecting own logs will most likely result in amplification and exploding of logs.

This is not accurate. Explosion of logs will happen only when using `logging` exporter that sends the logs to console (why is it not named `console` exporter is a mystery to me btw. 🤔).

Anyhow, I will rephrase this paragraph to warn more loudly about the logs explosion. 👍

rockb1017 replied:
if the file exporter would write to a path that it tails, only then, it will explode.

tigrannajaryan replied:
>This is not accurate. Explosion of logs will happen only when using logging exporter that sends the logs to console

You are right.

>if the file exporter would write to a path that it tails, only then, it will explode.

Correct. This is more difficult to do accidentally but also a danger. So, it would be good to warn about this too.

astencel-sumo replied:
I have added a "warning" section to the docs in https://github.com/open-telemetry/opentelemetry-helm-charts/pull/36/commits/f005dde14bfdd2971aad69a3c0adbe3233a78a97, please take a look.

astencel-sumo replied:
> Let's not advertise the `file` exporter yet. It is unstable and is typically used for local troubleshooting only. Can we show the example using some other exporter instead?

Changed to `otlphttp` exporter, please let me know if this is any better :)

tigrannajaryan replied:
LGTM.

dmitryax commented:
> @dmitryax any thoughts on what the implications of these change could be? Contrib is a much larger image, do we care about it?
The original idea for this chart was is to support only core collector functionality by default and provide users with an option to use the contrib collector for other use cases. I suppose that log collection components can be moved to the core collector build later, so I'm good with temporarily switching it to contrib.

naseemkullah commented:
> > @dmitryax any thoughts on what the implications of these change could be? Contrib is a much larger image, do we care about it?
> 
> The original idea for this chart was is to support only the core collector functionality by default and provide users with an option to use the contrib collector for other use cases. I suppose that log collection components can be moved to the core collector build later, so I'm good with temporarily switching it to contrib.

According to https://github.com/open-telemetry/opentelemetry-helm-charts/pull/36#discussion_r592389265 the `filelog` receiver will not be making it to core anytime soon.

I have a slight preference for sticking to the core image by default and just adding a sample values.yaml in the readme as how to achieve the container logging, but don't let that be a blocker!

astencel-sumo commented:
Folks, I have applied all the code review remarks:

- Switched to setting `agentCollector.containerLogs.enabled` to `false` by default,
- Switched back to using core collector image in the chart,
- Added docs on how to enable container log collection,
- Added docs on the warning of looping logs from exporter back into receiver,
- Changed `file` exporter in the example to `otlphttp` exporter, according to code review remarks.

Please take a look.

tigrannajaryan commented on charts/opentelemetry-collector/README.md:
```diff
@@ -107,6 +108,73 @@ agentCollector:
+To prevent this, the default configuration of the receiver excludes logs from the collector's containers.
+
+If you want to include the collector's logs, make sure to replace the `logging` exporter
+with an exporter that does not send logs to collector's standard output.
```
Please also add warning that `--log-level=debug` command line option should be avoided when collecting Collector's own logs since at debug level the Collector may produce logs per each collected input log item, which again can result in amplified loop.

astencel-sumo replied:
Added note in https://github.com/astencel-sumo/opentelemetry-helm-charts/commit/8a123a1ec05bf4197fa2f8edc89bba4dd1b810e2.

dmitryax commented on charts/opentelemetry-collector/values.yaml:
```diff
@@ -73,7 +80,6 @@ imagePullSecrets: []
 # OpenTelemetry Collector executable
 command:
   name: otelcol
-  # name: otelcontribcol
```
Can we bring back the comments? they seem to be useful

tigrannajaryan replied:
If we want comments let's make sure they are actual readable sentences and don't look like a commented code that we forgot to delete. 

A comment like the one that was on line 76 is not very helpful, it is confusing. I don't know if it is supposed to be uncommented or deleted. How do I know it should replace the previous line?

tigrannajaryan replied:
@dmitryax how do you want to move forward? Address this in this PR or a separate PR?

dmitryax replied:
Adding a comment would be helpful. I'm good with either way: addressing it here or in separate PR.

tigrannajaryan replied:
@astencel-sumo up to you. I can merge now or merge for you to address it.

astencel-sumo replied:
I'll add a comment to this.

astencel-sumo replied:
Added comment in https://github.com/open-telemetry/opentelemetry-helm-charts/pull/36/commits/04c8f4734cf4a5a48fbab40bc1ed8b6ba85da618.

dmitryax reviewed (commented):
One nit, otherwise LGTM

rockb1017 commented on charts/opentelemetry-collector/templates/_config.tpl:
```diff
@@ -115,9 +116,90 @@ exporters:
+receivers:
+  filelog:
+    include: [ /var/log/pods/*/*/*.log ]
+    # Exclude collector container's logs. The file format is /var/log/pods/<namespace_name>_<pod_name>_<pod_uid>/<container_name>/<run_id>.log
```
I think it should be configurable from values.yaml. 
k8s admin would may like to ingest otel agents' logs to monitor them after they disable `logging` exporter.

astencel-sumo replied:
Let's make it a separate PR if that's OK?

We probably need to prevent a scenario when user changes the property and inadvertently causes the looping of logs from exporter back into the receiver. It needs to be assured that changing of such property alone does not cause the loop.

tigrannajaryan replied:
Sounds good, let's do in a separate PR.

astencel-sumo commented:
I have hopefully addressed all the comments. Are we ready to merge @tigrannajaryan?

tigrannajaryan commented:
Thanks @astencel-sumo
